File: training-data/nim/rabin_karp_search.nim
proc rabinKarpSearch(text, pattern: string): seq[int] =
  let n = text.len
  let m = pattern.len
  if m == 0 or m > n:
    return @[]
  const base = 256
  const modulus = 101
  var patternHash = 0
  var windowHash = 0
  var h = 1
  for i in 0 ..< m - 1:
    h = (h * base) mod modulus
  for i in 0 ..< m:
    patternHash = (base * patternHash + ord(pattern[i])) mod modulus
    windowHash = (base * windowHash + ord(text[i])) mod modulus
  for i in 0 .. n - m:
    if patternHash == windowHash and text[i ..< i + m] == pattern:
      result.add(i)
    if i < n - m:
      windowHash = (base * (windowHash - ord(text[i]) * h) + ord(text[i + m])) mod modulus
      if windowHash < 0:
        windowHash += modulus

echo rabinKarpSearch("abxabcabcaby", "abcaby")
echo rabinKarpSearch("aaaaa", "aa")

File: training-data/kotlin/RabinKarpSearch.kt
fun rabinKarpSearch(text: String, pattern: String): List<Int> {
    if (pattern.isEmpty() || pattern.length > text.length) return emptyList()

    val base = 256
    val modulus = 1_000_000_007L
    val m = pattern.length
    var patternHash = 0L
    var windowHash = 0L
    var highOrder = 1L
    repeat(m - 1) { highOrder = (highOrder * base) % modulus }

    for (i in 0 until m) {
        patternHash = (patternHash * base + pattern[i].code) % modulus
        windowHash = (windowHash * base + text[i].code) % modulus
    }

    val matches = mutableListOf<Int>()
    for (i in 0..text.length - m) {
        if (patternHash == windowHash && text.substring(i, i + m) == pattern) {
            matches.add(i)
        }
        if (i < text.length - m) {
            windowHash = (base * (windowHash - text[i].code * highOrder) + text[i + m].code) % modulus
            if (windowHash < 0) windowHash += modulus
        }
    }
    return matches
}

fun main() {
    println(rabinKarpSearch("abxabcabcaby", "abcaby"))
    println(rabinKarpSearch("aaaaa", "aa"))
}
